File: 2015/03/go/main.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
)

type coordinate struct {
	x int
	y int
}

const (
	UP    = "^"
	RIGHT = ">"
	DOWN  = "v"
	LEFT  = "<"
)

func main() {
	fileName := filepath.Join("..", "inputs.txt")
	file, err := os.Open(fileName)

	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	houseGrid := createHouseGrid()

	santaPosition := coordinate{4, 4}
	robotSantaPosition := coordinate{4, 4}
	direction := make([]byte, 1)
	visitedHouses := 1
	isSantaTurn := true
	for _, err := file.Read(direction); err != io.EOF; _, err = file.Read(direction) {
		if isSantaTurn {
			houseGrid, robotSantaPosition = expandGrid(
				houseGrid,
				string(direction),
				santaPosition,
				robotSantaPosition,
			)
			visitedHouses += walkDirections(string(direction), houseGrid, &santaPosition)

			isSantaTurn = false
			continue
		}

		houseGrid, santaPosition = expandGrid(
			houseGrid,
			string(direction),
			robotSantaPosition,
			santaPosition,
		)
		visitedHouses += walkDirections(string(direction), houseGrid, &robotSantaPosition)
		isSantaTurn = true
	}

	fmt.Println(visitedHouses)
}

func createHouseGrid() [][]bool {
	houseGrid := make([][]bool, 10)

	for i := range houseGrid {
		houseGrid[i] = make([]bool, 10)
		// setting start position to the middle of the matrix
		if i == 4 {
			houseGrid[4][4] = true
		}
	}

	return houseGrid
}

func expandGrid(
	houseGrid [][]bool,
	direction string,
	primaryPosition,
	secondaryPosition coordinate,
) ([][]bool, coordinate) {
	switch direction {
	case UP:
		if primaryPosition.y-1 < 0 {
			temp := expandY(houseGrid)
			for i := range temp {
				for j := range temp[i] {
					if i > 0 {
						temp[i][j] = houseGrid[i-1][j]
					}
				}
			}
			secondaryPosition.y += 1
			return temp, secondaryPosition
		}
		return houseGrid, secondaryPosition
	case RIGHT:
		if primaryPosition.x+1 >= len(houseGrid[0]) {
			// add a new collumn to the end of each line
			temp := expandX(houseGrid)
			copyGridFromStart(temp, houseGrid)
			return temp, secondaryPosition
		}
		return houseGrid, secondaryPosition
	case DOWN:
		if primaryPosition.y+1 >= len(houseGrid) {
			// append a new line to the matrix;
			temp := expandY(houseGrid)
			copyGridFromStart(temp, houseGrid)
			return temp, secondaryPosition
		}
		return houseGrid, secondaryPosition
	case LEFT:
		if primaryPosition.x-1 < 0 {
			// add a new collumn to beginning of each line
			temp := expandX(houseGrid)
			for i := range houseGrid {
				for j := range houseGrid[i] {
					temp[i][j+1] = houseGrid[i][j]
				}
			}
			secondaryPosition.x += 1
			return temp, secondaryPosition
		}
		return houseGrid, secondaryPosition
	default:
		return houseGrid, secondaryPosition
	}
}

func expandY(houseGrid [][]bool) [][]bool {
	temp := make([][]bool, len(houseGrid)+1)
	for i := range temp {
		temp[i] = make([]bool, len(houseGrid[0]))
	}
	return temp
}

func expandX(houseGrid [][]bool) [][]bool {
	temp := make([][]bool, len(houseGrid))
	for i := range temp {
		temp[i] = make([]bool, len(houseGrid[0])+1)
	}
	return temp
}

func copyGridFromStart(newGrid, oldGrid [][]bool) {
	for y := range oldGrid {
		for x := range oldGrid[y] {
			newGrid[y][x] = oldGrid[y][x]
		}
	}
}

func walkDirections(
	direction string,
	houseGrid [][]bool,
	primaryPosition *coordinate,
) int {
	switch direction {
	case UP:
		if primaryPosition.y-1 >= 0 {
			primaryPosition.y -= 1
		}
		return checkVisitedHouse(houseGrid, *primaryPosition)
	case RIGHT:
		if primaryPosition.x+1 < len(houseGrid[0]) {
			primaryPosition.x += 1
		}
		return checkVisitedHouse(houseGrid, *primaryPosition)
	case DOWN:
		if primaryPosition.y+1 < len(houseGrid) {
			primaryPosition.y += 1
		}
		return checkVisitedHouse(houseGrid, *primaryPosition)
	case LEFT:
		if primaryPosition.x-1 >= 0 {
			primaryPosition.x -= 1
		}
		return checkVisitedHouse(houseGrid, *primaryPosition)
	default:
		return 0
	}
}

func checkVisitedHouse(houseGrid [][]bool, position coordinate) int {
	if !houseGrid[position.y][position.x] {
		houseGrid[position.y][position.x] = true
		return 1
	}

	return 0
}

File: 2015/03/python/main.py
from os.path import join
from enum import Enum


class Direction(Enum):
    UP = "^"
    RIGHT = ">"
    DOWN = "v"
    LEFT = "<"


def calculate_visited_houses(
        house_grid: list[list[bool]],
        position: dict,
        secondary_position: dict
) -> int:
    first_row_len = len(house_grid[position["y"]])

    match character:
        case Direction.UP.value:
            if (position["y"] - 1 < 0):
                new_line = [False for x in range(first_row_len)]
                house_grid.insert(0, new_line)
                position["y"] += 1
                secondary_position["y"] += 1
            position["y"] -= 1

            return check_new_house(house_grid, position)
        case Direction.RIGHT.value:
            if (position["x"] + 1 >= first_row_len):
                for row in house_grid:
                    row.append(False)
            position["x"] += 1

            return check_new_house(house_grid, position)
        case Direction.DOWN.value:
            if (position["y"] + 1 >= len(house_grid)):
                new_line = [False for x in range(first_row_len)]
                house_grid.append(new_line)
            position["y"] += 1

            return check_new_house(house_grid, position)
        case Direction.LEFT.value:
            if (position["x"] - 1 < 0):
                for row in house_grid:
                    row.insert(0, False)
                position["x"] += 1
                secondary_position["x"] += 1
            position["x"] -= 1

            return check_new_house(house_grid, position)

    return 0


def check_new_house(house_grid: list[list[bool]], position: dict) -> int:
    delivered_presents = 0
    if (not house_grid[position["y"]][position["x"]]):
        delivered_presents = 1
    house_grid[position["y"]][position["x"]] = True
    return delivered_presents


if __name__ == "__main__":
    file_name = join("..", "inputs.txt")

    position = {"x": 4, "y": 4}
    robot_position = {"x": 4, "y": 4}
    house_grid = [[False for x in range(10)] for y in range(10)]
    house_grid[position["x"]][position["y"]] = True
    count = 1
    is_santa_turn = True
    with open(file_name) as f:
        while True:
            character = f.read(1)
            if (character == ""):
                break

            if (is_santa_turn):
                count += calculate_visited_houses(
                    house_grid,
                    position,
                    robot_position
                )
                is_santa_turn = False
                continue

            count += calculate_visited_houses(
                house_grid,
                robot_position,
                position
            )
            is_santa_turn = True

    print(count)
